packages/core/src/DataSchema.ts

```diff
@@ -0,0 +1,118 @@
+import { CircuitConfig } from '@unirep/circuits'
+
+export class DataSchema {
+    public schema: any[]
+    public config: CircuitConfig
+    public data: bigint[]
+
+    constructor(schema, config = CircuitConfig.default) {
+        this.config = config
+        this.schema = this.parseSchema(schema)
+        this.data = new Array(config.FIELD_COUNT).fill(BigInt(0))
+    }
+
+    parseSchema(schema: any[]): any[] {
+        let sumOffset = 0
+        let replOffset = 0
+        const maxSumOffset = 253 * this.config.SUM_FIELD_COUNT
+        const maxReplOffset =
+            253 * (this.config.FIELD_COUNT - this.config.SUM_FIELD_COUNT)
+
+        return schema.map((field, idx) => {
+            const { name, type, updateBy, ...extranousFields } = field
+            const match = type.match(/^uint(\d+)$/)
+            if (match === null)
+                throw new Error(`Invalid type for field ${name}: "${type}"`)
+            if (Object.keys(extranousFields).length !== 0) {
+                throw new Error(
+                    `Invalid fields included for field ${name}: [${Object.keys(
+                        extranousFields
+                    )}]`
+                )
+            }
+
+            const duplicateEntry = schema.find(
+                (x, i) => x.name == field.name && idx != i
+            )
+            if (duplicateEntry) {
+                throw Error(
+                    `Schema includes a duplicate entry: "${duplicateEntry.name}"`
+                )
+            }
+
+            const bits = +match[1]
+            if (bits < 1 || bits > 253)
+                throw new Error(`Invalid uint size for field ${name}: ${bits}`)
+            if (updateBy === 'sum') {
+                if (
+                    Math.floor(sumOffset / 253) !==
+                    Math.floor((sumOffset + bits - 1) / 253)
+                ) {
+                    sumOffset += 253 - (sumOffset % 253)
+                }
+                if (sumOffset + bits > maxSumOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+                const dataIndex = Math.floor(sumOffset / 253)
+                const offset = sumOffset % 253
+                sumOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            } else if (updateBy === 'replace') {
+                if (bits !== 253 - this.config.REPL_NONCE_BITS)
+                    throw new Error(
+                        `Invalid uint size for field ${name}: ${bits}`
```

Comment:
Maybe in the error message add a note saying `field must be ${253 - this.config.REPL_NONCE_BITS} bits`

Reply:
Done, using this error message now